### PR TITLE
add `immutable-vector`, `immutable-vector-copy`, etc.

### PR DESCRIPTION
**Pull request by mflatt**
Based on a suggestion and initial work from @owaddell, extract the `immutable-vector` addition from #596, and then add `immutable-vector-copy` and friends. The motivation for the additions is the same as for `vector-copy` and company: to implement persistent data structures, but where you want to enforce immutability and maybe enable optimizations.

The changes include cp0 rules to combine vector-construction operations, such as turning `(vector->immutable-vector (vector-append (vector x y) '#(3)))` into `(immutable-vector x y 3)`.

It turns out that `(immutable-vector-copy x)` and `(immutable-vector-append x)` are both equivalent to `(vector->immutable-vector x)`. This PR doesn't try to prune the API, although it's possible that we would have left out `vector->immutable-vector` if `immutable-vector-copy` existed in the first place.

### CODE REVIEW

**Review by burgerrg — APPROVED** [review]
These procedures make immutable vectors much easier to use, thanks!

**Comment by owaddell** [comment]
I'd like to extend the inline handler for `vector-append` and `immutable-vector-append` slightly to handle the following:
``` scheme
(define (expand-segments segments)
  (let ([n (vector-length segments)])
    (immutable-vector-append segments (make-vector n #f))))
```
Right now cptypes doesn't recognize that `n` will be a safe argument for `make-vector`, so it doesn't switch to `#3%make-vector`. @gus-massa are you interested in helping refine cptypes to handle this?

Current work-in-progress patch to cpprims.ss: [patch.txt](https://github.com/cisco/ChezScheme/files/13865194/patch.txt)

**Comment by mflatt** [comment]
This PR now pulls in @owaddell's improvements. There's no cptypes changes, yet, and those might be best considered in a new PR.

File: csug/objects.stex [hunk 1]
```diff
@@ -920,6 +922,22 @@ is immutable; otherwise, the result is an immutable vector with the same content
...
+\formdef{immutable-vector-copy}{\categoryprocedure}{(immutable-vector-copy \var{vector})}
+\formdef{immutable-vector-copy}{\categoryprocedure}{(immutable-vector-copy \var{vector} \var{start} \var{n})}
+\formdef{immutable-vector-append}{\categoryprocedure}{(immutable-vector-append \var{vector} \dots)}
+\formdef{immutable-vector-set/copy}{\categoryprocedure}{(vector-set/copy \var{vector} \var{n} \var{val})}
+\listlibraries
+\endentryheader
+
+Like \scheme{vector}, \scheme{vector-copy},  \scheme{vector-append}, and
+\scheme{vector-set/copy}, but producing an immutable vector instead of a mutable
+vector. In the case of \scheme{immutable-vector-copy},  \scheme{immutable-vector-append}, or
```
[the remark this answers is not in the document]
Comment (mflatt) [reply csug/objects.stex]: Fixed.

File: csug/objects.stex [hunk 2]
```diff
@@ -920,6 +922,22 @@ is immutable; otherwise, the result is an immutable vector with the same content
...
+\formdef{immutable-vector}{\categoryprocedure}{(immutable-vector \var{obj} \dots)}
+\formdef{immutable-vector-copy}{\categoryprocedure}{(immutable-vector-copy \var{vector})}
+\formdef{immutable-vector-copy}{\categoryprocedure}{(immutable-vector-copy \var{vector} \var{start} \var{n})}
+\formdef{immutable-vector-append}{\categoryprocedure}{(immutable-vector-append \var{vector} \dots)}
+\formdef{immutable-vector-set/copy}{\categoryprocedure}{(vector-set/copy \var{vector} \var{n} \var{val})}
+\listlibraries
+\endentryheader
+
+Like \scheme{vector}, \scheme{vector-copy},  \scheme{vector-append}, and
+\scheme{vector-set/copy}, but producing an immutable vector instead of a mutable
```
[the remark this answers is not in the document]
Comment (mflatt) [reply csug/objects.stex]: Thanks - rephrased. I also moved the part about potentially returning an existing object to this sentence.

**Comment by mflatt** [comment]
More improvements from @owaddell now added.

**Pull request merged**